### DISCUSSION TITLE
Building large React Native apps with Bazel

### DISCUSSION THREAD

**Discussion by zachgrayio** [MobileNativeFoundation/discussions#145] [General]
Hey folks!

We've been chatting with some members of the Bazel community about adding proper support for building large-scale React Native applications with Bazel, and I am interested in gauging interest in this topic with this community as well.

Anyone who's looked into this at all will know there's quite a few challenges to getting up and running with RN+Bazel, and perhaps we'll enumerate those in this thread at some point but for now we'd like to focus on simply gauging interest & understanding if anyone has had any success with this internally and at scale.

1. Is your organization interested in using Bazel to build React Native applications?
2. Are you using Bazel to build RN applications today? If so, please provide a brief summary of the patches/hacks/workarounds necessary to get this done and I'll follow up with more specific questions
3. Are you and/or your organization willing to contribute to the development of first-class support for Bazel+RN?

Looking forward to responses from the MNF community--thanks!

**Comment by zachgrayio** [comment]
Small update:

Over at `rules_nodejs` the react native support [issue](https://github.com/bazelbuild/rules_nodejs/issues/2541) has been closed and the maintainers understandably don't want to take on this scope. With that in mind:

- As I mentioned in my comment over there, we're likely going to move ahead on assembling a working group to collaborate on this, ideally with contributions from [Flare](https://flare.build/) in addition to a few other organizations who are interested in making this a reality. Please comment here and/or reach out to me if you'd like to get involved.
- Looks like we'll be working on a ruleset over the coming months to implement this support, assuming this moves ahead. TBD on where this ends up living long-term, but I'm assuming that will be some place like the MNF, bazelbuild, or bazel-contrib org 

Also, some anecdotes from one of the few 'success' cases I've heard of Bazel + React Native:
- project was fairly small from what I gather, unsure on LOC, but small (will see if I can find this)
- that team is moving away from react native and thus didn't put a lot of effort into this and doesn't have interest in doing so in the future
- the main issues lie in the `metro` bundler which doesn't seem to cope well with the sandbox & symlinks
- workarounds in this case were mostly some large ugly genrules to essentially glob the whole source tree in the manner that metro expects, losing any benefit of using bazel here
- little / no integration with Bazel's ios & android support

**Reply by rockwotj** [reply]
Hey Zach! To add a voice here in the 'success' case - we prebuild the native side of react native into an xcframework (or use the prebuilt jar from walmart labs [here](https://mvnrepository.com/artifact/com.walmartlabs.ern/react-native)). This is a good tradeoff for us as React Native is really annoying to build.

For release builds we are also just glob the entire source tree for `metro` to build in a temporary directory. It mostly is working well for us as the DX of HMR `metro` gives is awesome for development, and release builds aren't painful (yet) at tens of thousands of lines of code.

Another caveat is that we never really depend on any third party native modules - we have found the majority in the ecosystem to be very low quality or simple enough that we just develop our own.

I'd love to hear if there has been progress on using this in Bazel more. I've always wondered what Facebook does internally with BUCK. I know they have some BUCK files in the open source repo, but there is no instructions on how to actually use BUCK, let alone port something similar to Bazel.

**Reply by farcaller** [reply]
the other large issue is gradle that isn't straightforward to wrap into bazel rules either (for android builds).
